File: reference/rtp.py
"""Reference implementation of the RTP/JPEG (RFC 2435), L16 (RFC 3551) and RTCP SR logic.
The Rust port (src/protocols/rtp.rs, src/jpeg.rs) mirrors this line for line."""
import json, struct
STD = {k: v for k, v in json.load(open("std_huff.json")).items()}
STD_BY_ID = {(0,0): STD["DC_LUM"], (0,1): STD["DC_CHROMA"], (1,0): STD["AC_LUM"], (1,1): STD["AC_CHROMA"]}

def analyze_jpeg(d):
    if len(d) < 4 or d[0] != 0xFF or d[1] != 0xD8: raise ValueError("not a JPEG")
    qt = {}; comps = []; w = h = 0; dri = 0; custom = False; sof = None
    pos = 2; scan_start = None
    while pos + 4 <= len(d):
        if d[pos] != 0xFF: raise ValueError("bad marker at %d" % pos)
        m = d[pos+1]
        if m == 0xFF: pos += 1; continue
        if m == 0xD8 or m == 0x01 or 0xD0 <= m <= 0xD7: pos += 2; continue
        ln = (d[pos+2] << 8) | d[pos+3]
        if ln < 2 or pos + 2 + ln > len(d): raise ValueError("truncated segment")
        seg = d[pos+4:pos+2+ln]
        if m == 0xDB:
            o = 0
            while o < len(seg):
                pq, tq = seg[o] >> 4, seg[o] & 15; o += 1
                if pq != 0: raise ValueError("16-bit quant tables unsupported")
                if o + 64 > len(seg) or tq > 3: raise ValueError("bad DQT")
                qt[tq] = bytes(seg[o:o+64]); o += 64
        elif m == 0xC0:
            if len(seg) < 6 or seg[0] != 8: raise ValueError("not 8-bit baseline")
            h = (seg[1] << 8) | seg[2]; w = (seg[3] << 8) | seg[4]; n = seg[5]
            if n != 3 or len(seg) < 6 + 3*n: raise ValueError("need 3 components")
            comps = [(seg[6+3*i], seg[7+3*i], seg[8+3*i]) for i in range(3)]
            sof = m
        elif m in (0xC1, 0xC2, 0xC3, 0xC5, 0xC6, 0xC7, 0xC9, 0xCA, 0xCB):
            raise ValueError("only baseline JPEG can go over RTP/JPEG")
        elif m == 0xC4:
            o = 0
            while o + 17 <= len(seg):
                tc, th = seg[o] >> 4, seg[o] & 15
                bits = list(seg[o+1:o+17]); n = sum(bits); vals = list(seg[o+17:o+17+n]); o += 17 + n
                std = STD_BY_ID.get((tc, th))
                if std is None or std[0] != bits or std[1] != vals: custom = True
        elif m == 0xDD:
            dri = (seg[0] << 8) | seg[1]
        elif m == 0xDA:
            scan_start = pos + 2 + ln; break
        pos += 2 + ln
    if sof is None or scan_start is None: raise ValueError("no SOF/SOS")
    hv = [c[1] for c in comps]
    if hv[1] != 0x11 or hv[2] != 0x11: raise ValueError("chroma must be 1x1")
    if hv[0] == 0x21: typ = 0
    elif hv[0] == 0x22: typ = 1
    else: raise ValueError("unsupported sampling %02x" % hv[0])
    tq0, tq1 = comps[0][2], comps[1][2]
    if comps[2][2] != tq1: raise ValueError("Cb/Cr use different quant tables")
    if tq0 not in qt or tq1 not in qt: raise ValueError("missing quant table")
    if w == 0 or h == 0 or (w + 7) // 8 > 255 or (h + 7) // 8 > 255: raise ValueError("size unsupported")
    end = len(d)
    while end >= 2 and not (d[end-2] == 0xFF and d[end-1] == 0xD9): end -= 1
    scan_end = end - 2 if end >= 2 else len(d)
    if scan_end <= scan_start: raise ValueError("empty scan")
    return dict(width=w, height=h, typ=typ + (64 if dri else 0), qtables=qt[tq0] + qt[tq1], dri=dri,
                scan_start=scan_start, scan_end=scan_end, custom_huffman=custom)

def rtp_header(marker, pt, seq, ts, ssrc):
    return struct.pack(">BBHII", 0x80, (0x80 if marker else 0) | pt, seq & 0xFFFF, ts & 0xFFFFFFFF, ssrc)

def packetize_jpeg(info, data, ts, ssrc, seq, max_payload=1400):
    scan = data[info["scan_start"]:info["scan_end"]]
    pkts = []; off = 0
    while True:
        first = off == 0
        hdr = bytes([0, (off >> 16) & 0xFF, (off >> 8) & 0xFF, off & 0xFF, info["typ"], 255,
                     (info["width"] + 7) // 8, (info["height"] + 7) // 8])
        if info["typ"] >= 64: hdr += struct.pack(">HH", info["dri"], 0xFFFF)
        if first: hdr += bytes([0, 0]) + struct.pack(">H", len(info["qtables"])) + info["qtables"]
        room = max_payload - len(hdr)
        chunk = scan[off:off+room]; off += len(chunk)
        last = off >= len(scan)
        pkts.append(rtp_header(last, 26, seq, ts, ssrc) + hdr + chunk); seq = (seq + 1) & 0xFFFF
        if last: break
    return pkts, seq

def packetize_l16(pcm_le, chans, pos_frames, ssrc, seq, base_ts, pos0, max_payload=1200):
    """pcm_le: little-endian S16 interleaved. RTP timestamp = base_ts + (pos - pos0) in sample frames."""
    fb = 2 * chans; per = max(1, max_payload // fb); pkts = []
    n_frames = len(pcm_le) // fb; i = 0
    while i < n_frames:
        k = min(per, n_frames - i)
        raw = pcm_le[i*fb:(i+k)*fb]
        be = bytearray(raw); be[0::2], be[1::2] = raw[1::2], raw[0::2]
        pkts.append(rtp_header(False, 97, seq, (base_ts + (pos_frames - pos0) + i) & 0xFFFFFFFF, ssrc) + bytes(be))
        seq = (seq + 1) & 0xFFFF; i += k
    return pkts, seq

def ntp_from_unix_micros(us):
    sec = us // 1_000_000 + 2208988800; frac = ((us % 1_000_000) << 32) // 1_000_000
    return sec & 0xFFFFFFFF, frac & 0xFFFFFFFF

def sender_report(ssrc, unix_us, rtp_ts, pkt_count, octet_count, cname=b"uvcweb"):
    sec, frac = ntp_from_unix_micros(unix_us)
    sr = struct.pack(">BBHIIIIII", 0x80, 200, 6, ssrc, sec, frac, rtp_ts & 0xFFFFFFFF, pkt_count & 0xFFFFFFFF, octet_count & 0xFFFFFFFF)
    item = bytes([1, len(cname)]) + cname + b"\0"
    while (len(item) + 4) % 4: item += b"\0"          # SDES chunk (ssrc + items) padded to 32 bit
    sdes_len = (4 + 4 + len(item)) // 4 - 1
    sdes = struct.pack(">BBHI", 0x81, 202, sdes_len, ssrc) + item
    return sr + sdes

File: src/rtp.rs
//! RTP / RTCP packet building shared by every RTP-based protocol (RTSP today).
//! Byte-for-byte port of the ffmpeg-verified Python reference (proto/rtp.py);
//! `golden_tests.rs` checks that they agree.

use crate::jpeg::JpegInfo;
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::{SystemTime, UNIX_EPOCH};

pub const PT_JPEG: u8 = 26; // static payload type for RTP/JPEG
pub const PT_L16: u8 = 97; // dynamic payload type used for 16-bit PCM

static RNG: AtomicU64 = AtomicU64::new(0x9E37_79B9_7F4A_7C15);

/// Good-enough random numbers for SSRCs, sequence numbers and session ids (no rand crate).
pub fn rand_u32() -> u32 {
    let t = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos() as u64)
        .unwrap_or(1);
    let mut x = RNG.fetch_add(0x9E37_79B9_7F4A_7C15, Ordering::Relaxed) ^ t;
    x ^= x >> 30;
    x = x.wrapping_mul(0xBF58_476D_1CE4_E5B9);
    x ^= x >> 27;
    x = x.wrapping_mul(0x94D0_49BB_1331_11EB);
    x ^= x >> 31;
    x as u32
}

pub fn rtp_header(out: &mut Vec<u8>, marker: bool, pt: u8, seq: u16, ts: u32, ssrc: u32) {
    out.push(0x80); // version 2, no padding, no extension, no CSRC
    out.push(if marker { 0x80 | pt } else { pt });
    out.extend_from_slice(&seq.to_be_bytes());
    out.extend_from_slice(&ts.to_be_bytes());
    out.extend_from_slice(&ssrc.to_be_bytes());
}

/// Split one JPEG picture into RFC 2435 packets. `max_payload` is the RTP payload budget
/// (JPEG headers included). Quantisation tables travel in the first packet of every frame (Q=255).
pub fn packetize_jpeg(
    info: &JpegInfo,
    data: &[u8],
    ts: u32,
    ssrc: u32,
    seq: &mut u16,
    max_payload: usize,
) -> Vec<Vec<u8>> {
    let scan = &data[info.scan_start..info.scan_end];
    let w8 = (info.width as usize).div_ceil(8) as u8;
    let h8 = (info.height as usize).div_ceil(8) as u8;
    let mut out: Vec<Vec<u8>> = Vec::new();
    let mut off = 0usize;
    loop {
        let mut hdr: Vec<u8> = vec![
            0,
            (off >> 16) as u8,
            (off >> 8) as u8,
            off as u8,
            info.rtp_type,
            255,
            w8,
            h8,
        ];
        if info.rtp_type >= 64 {
            hdr.extend_from_slice(&info.dri.to_be_bytes());
            hdr.extend_from_slice(&[0xFF, 0xFF]); // F=1, L=1, count=0x3FFF: not aligned to restart intervals
        }
        if off == 0 {
            hdr.extend_from_slice(&[0, 0]); // MBZ, precision (all tables 8 bit)
            hdr.extend_from_slice(&(info.qtables.len() as u16).to_be_bytes());
            hdr.extend_from_slice(&info.qtables);
        }
        let room = max_payload.saturating_sub(hdr.len()).max(1);
        let end = (off + room).min(scan.len());
        let last = end >= scan.len();
        let mut p: Vec<u8> = Vec::with_capacity(12 + hdr.len() + (end - off));
        rtp_header(&mut p, last, PT_JPEG, *seq, ts, ssrc);
        *seq = seq.wrapping_add(1);
        p.extend_from_slice(&hdr);
        p.extend_from_slice(&scan[off..end]);
        out.push(p);
        off = end;
        if last {
            break;
        }
    }
    out
}

/// Split interleaved little-endian S16 PCM into RTP L16 packets (network byte order, RFC 3551).
/// `pos` / `pos0` are running sample-frame counters; the RTP timestamp is `base_ts + (pos - pos0)`.
#[allow(clippy::too_many_arguments)]
pub fn packetize_l16(
    pcm_le: &[u8],
    chans: usize,
    pos: u64,
    pos0: u64,
    base_ts: u32,
    ssrc: u32,
    seq: &mut u16,
    max_payload: usize,
) -> Vec<Vec<u8>> {
    let fb = 2 * chans.max(1);
    let per = (max_payload / fb).max(1);
    let n_frames = pcm_le.len() / fb;
    let mut out: Vec<Vec<u8>> = Vec::new();
    let mut i = 0usize;
    while i < n_frames {
        let k = per.min(n_frames - i);
        let ts = base_ts.wrapping_add((pos - pos0 + i as u64) as u32);
        let mut p: Vec<u8> = Vec::with_capacity(12 + k * fb);
        rtp_header(&mut p, false, PT_L16, *seq, ts, ssrc);
        *seq = seq.wrapping_add(1);
        for s in pcm_le[i * fb..(i + k) * fb].as_chunks::<2>().0 {
            p.push(s[1]); // swap to big endian
            p.push(s[0]);
        }
        out.push(p);
        i += k;
    }
    out
}

/// NTP timestamp (seconds since 1900, 32.32 fixed point) for a Unix time in microseconds.
pub fn ntp_from_unix_micros(us: i64) -> (u32, u32) {
    let sec = us.div_euclid(1_000_000) + 2_208_988_800;
    let rem = us.rem_euclid(1_000_000) as u64;
    let frac = (rem << 32) / 1_000_000;
    (sec as u32, frac as u32)
}

/// RTCP sender report followed by an SDES CNAME chunk (a valid compound packet).
/// It tells receivers which RTP timestamp corresponds to which wall-clock time, which is
/// how players line the audio and video streams up.
pub fn sender_report(
    ssrc: u32,
    unix_us: i64,
    rtp_ts: u32,
    pkt_count: u32,
    octet_count: u32,
    cname: &str,
) -> Vec<u8> {
    let (sec, frac) = ntp_from_unix_micros(unix_us);
    let mut p: Vec<u8> = Vec::with_capacity(48);
    p.extend_from_slice(&[0x80, 200, 0, 6]);
    p.extend_from_slice(&ssrc.to_be_bytes());
    p.extend_from_slice(&sec.to_be_bytes());
    p.extend_from_slice(&frac.to_be_bytes());
    p.extend_from_slice(&rtp_ts.to_be_bytes());
    p.extend_from_slice(&pkt_count.to_be_bytes());
    p.extend_from_slice(&octet_count.to_be_bytes());
    // SDES: one chunk (ssrc + CNAME item + terminator), padded to a 32-bit boundary
    let mut item: Vec<u8> = vec![1, cname.len() as u8];
    item.extend_from_slice(cname.as_bytes());
    item.push(0);
    while !(item.len() + 4).is_multiple_of(4) {
        item.push(0);
    }
    let words = (4 + 4 + item.len()) / 4 - 1;
    p.extend_from_slice(&[0x81, 202]);
    p.extend_from_slice(&(words as u16).to_be_bytes());
    p.extend_from_slice(&ssrc.to_be_bytes());
    p.extend_from_slice(&item);
    p
}
